### PR TITLE
Enable manual batch slot management for `tt_transformers`

### PR DESCRIPTION
**Pull request by rdraskicTT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/23512

### Problem description
When running models in data-parallel configuration it could happen that certain users get incorrectly scheduled for execution on a different model instance, which in turn causes accuracy drops. This PR introduces a way to properly manage user slots both during prefill and decode. Will be enabled in vllm through https://github.com/tenstorrent/vllm/pull/135

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/16422992325) CI passes
- [x] [DP=4 evals](https://github.com/tenstorrent/tt-shield/actions/runs/16423480191)
- [x] [DP=4 benchmark](https://github.com/tenstorrent/tt-shield/actions/runs/16423471165)
- [x] [TG demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/runs/16426096007) CI passes
- [ ] [T3K unit+demo](https://github.com/tenstorrent/tt-metal/actions/runs/16426076617)
- [x] [vllm nightly](https://github.com/tenstorrent/tt-metal/actions/runs/16490486791)

### CODE REVIEW

File: models/tt_transformers/tt/generator.py [hunk 1]
```diff
@@ -56,67 +56,56 @@ def __init__(self, model, model_args, mesh_device, tokenizer=None, formatter=Non
...
         batch_size, batch_seq_len = tokens.shape
-        max_batch_size_per_model, batch_size_per_model = self._get_batch_size_per_model(batch_size)
+        max_batch_size_per_model = self.model_args[0].max_batch_size
 
         # Each model expected to run the same model, safe to use 1st vocab size
         output_logits = torch.zeros(batch_size, 1, self.model_args[0].vocab_size)
         prompt_lens = prompt_lens if prompt_lens is not None else torch.tensor([batch_seq_len] * batch_size)
 
-        if page_table is not None:
-            assert isinstance(page_table, torch.Tensor), "page_table mush be torch.Tensor"
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py:68]: Can this assert be kept?

File: models/tt_transformers/tt/generator.py [hunk 2]
```diff
@@ -56,67 +56,56 @@ def __init__(self, model, model_args, mesh_device, tokenizer=None, formatter=Non
...
+            )
+            if page_table is not None:
+                page_table_user = self._get_prefill_user_page_table(
+                    page_table[idx : idx + 1], kv_cache[model_id], seq_len
                 )
 
+            logits = self.prefill_forward_single_user_text(
+                prefill_ids,
+                page_table=page_table_user if page_table is not None else None,
+                user_id=0 if page_table is not None else group_user_id,
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: Just to make sure I understand correctly, previously the page table was of size `max_batch_size_per_model` and was indexed by `group_user_id` in `paged_fill_cache` in attention, but now it is directly indexed here and it's passed into attention with size 1 and `user_id=0` (and we make sure the DP group is correct by determining `model_id` based on `user_id` from `empty_slots`) right? I think this should be fine, but just checking. @cglagovich this look ok to you too?

File: models/tt_transformers/tt/generator.py [hunk 3]
```diff
@@ -499,82 +488,77 @@ def prefill_forward(
...
         decode_output_xattn_masks = [None for _ in range(batch_size)]
         decode_output_full_text_row_masked_out_masks = [None for _ in range(batch_size)]
 
-        if page_table is not None:
-            assert isinstance(page_table, torch.Tensor), "page_table mush be torch.Tensor"
-            page_table = torch.split(page_table, batch_size_per_model)
-
-        if cross_page_table is not None:
-            assert isinstance(cross_page_table, torch.Tensor), "cross_page_table mush be torch.Tensor"
-            cross_page_table = torch.split(cross_page_table, batch_size_per_model)
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: Can the asserts be kept?

File: models/tt_transformers/tt/generator.py [hunk 4]
```diff
@@ -499,82 +488,77 @@ def prefill_forward(
...
+                page_table=user_page_table,
+                kv_cache=user_kv_cache,
+                cross_page_table=user_cross_page_table,
+                model_id=model_id,
+            )
+
+            if xattn_caches is not None:
+                xattn_caches[model_id] = xattn_cache
+
+            out_list[model_id][group_user_id] = logits
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: Any reason for this to be 2d or can it be 1d similar to prefill_forward_text?

File: models/tt_transformers/tt/generator.py [hunk 5]
```diff
@@ -656,19 +640,16 @@ def read_decode_output(self, tt_out, unpadded_batch, is_tokens=False):
...
-            if batch_size_per_model[i] == 0:
-                continue
             logits_i = self.model[i].process_output_decode(
-                tt_out[i], B=batch_size_per_model[i], S=1, is_tokens=is_tokens
+                tt_out[i], B=max_batch_size_per_model, S=1, is_tokens=is_tokens
             )
             logits.append(logits_i)
 
         logits = torch.cat(logits, 0)
-        assert logits.shape[0] == unpadded_batch
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py:671]: With this change, there should be a slice to `unpadded_batch` now right?

Comment (sraizada-tt) [reply models/tt_transformers/tt/generator.py:671]: no, it needs to be padded because vllm with permute this using the inverse perm table

Comment (skhorasganiTT) [reply models/tt_transformers/tt/generator.py:671]: Should unpadded_batch not be passed in then for the dp kv cache case so we don't have an unused arg? Also I think when sample on device is enabled in vLLM the unpadded_batch slice currently doesn't occur in vLLM (https://github.com/tenstorrent/vllm/blob/dev/vllm/worker/tt_model_runner.py#L776) so we should be careful not to break that.

Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py:671]: I had to keep this to be of `unpadded_batch` size since it was drastically affecting performance for more users.

File: models/tt_transformers/tt/generator.py (hunk 4 above)
Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py]: No, I just wanted to avoid refactoring more than I need to. If you think it's better to have them work in the same way I can address this.

File: models/tt_transformers/tt/generator.py (hunk 2 above)
Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py]: Yes, other rows were not used downstream. This actually caused a slight performance increase.

File: models/tt_transformers/tt/generator.py (hunk 4 above)
Comment (skhorasganiTT) [reply models/tt_transformers/tt/generator.py]: Yes I think they should be kept consistent if it's not much work

Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py]: Done, I did some further refactor to keep things shorter, please check

File: models/tt_transformers/tt/generator.py [hunk 6]
```diff
@@ -651,26 +637,33 @@ def decode_forward(
...
+        padded_batch_size = max_batch_size_per_model * self.data_parallel
+
+        if empty_slots is None:
+            empty_slots = list(range(unpadded_batch, padded_batch_size))
+
+        empty_slots_set = set(empty_slots)
+        mask = torch.tensor([i not in empty_slots_set for i in range(padded_batch_size)], dtype=torch.bool)
+
+        return torch.chunk(mask, self.data_parallel, 0)
+
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: this fcn still exists (unused)

File: models/tt_transformers/tt/generator.py [hunk 7]
```diff
@@ -651,26 +637,33 @@ def decode_forward(
...
     # Note: This function is called by vLLM
     def read_decode_output(self, tt_out, unpadded_batch, is_tokens=False):
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: unpadded_batch arg is now unused. I think for the sampling on device case in vLLM it expects to receive an unpadded batch so might need to modify vllm

File: models/tt_transformers/tt/generator.py (hunk 6 above)
Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py]: Removed

File: models/tt_transformers/tt/generator.py [hunk 8]
```diff
@@ -56,66 +56,56 @@ def __init__(self, model, model_args, mesh_device, tokenizer=None, formatter=Non
...
+            last_token_idx = seq_len - 1
 
-                seq_len = int(prompt_lens[user_id])
-                last_token_idx = seq_len - 1
+            prefill_seq_len = get_padded_prefill_len(seq_len)
+            prefill_ids = torch.cat(
+                [tokens[idx : idx + 1, :seq_len], torch.zeros(1, prefill_seq_len - seq_len).long()], dim=-1
+            )
+            if page_table is not None:
+                page_table_user = self._get_prefill_user_page_table(page_table[idx:], kv_cache[model_id], seq_len)
```
Comment (ppetrovicTT) [inline models/tt_transformers/tt/generator.py]: shouldn't we use `page_table[idx : idx + 1]`?

File: models/tt_transformers/tt/generator.py [hunk 9]
```diff
@@ -56,66 +56,56 @@ def __init__(self, model, model_args, mesh_device, tokenizer=None, formatter=Non
...
+            seq_len = int(prompt_lens[idx])
+            last_token_idx = seq_len - 1
 
-                seq_len = int(prompt_lens[user_id])
-                last_token_idx = seq_len - 1
+            prefill_seq_len = get_padded_prefill_len(seq_len)
+            prefill_ids = torch.cat(
+                [tokens[idx : idx + 1, :seq_len], torch.zeros(1, prefill_seq_len - seq_len).long()], dim=-1
+            )
+            if page_table is not None:
```
Comment (ppetrovicTT) [inline models/tt_transformers/tt/generator.py]: since page_table is being used only in this part, why don't we assert here instead of beginning of the function?
`isinstance(page_table, torch.Tensor)`

File: models/tt_transformers/tt/generator.py [hunk 10]
```diff
@@ -56,66 +56,56 @@ def __init__(self, model, model_args, mesh_device, tokenizer=None, formatter=Non
...
+                page_table_user = self._get_prefill_user_page_table(page_table[idx:], kv_cache[model_id], seq_len)
+
+            logits = self.prefill_forward_single_user_text(
+                prefill_ids,
+                page_table=page_table_user if page_table is not None else None,
```
Comment (ppetrovicTT) [inline models/tt_transformers/tt/generator.py]: page_table is not None! we have it checked just above

File: models/tt_transformers/tt/generator.py (hunk 9 above)
Comment (ppetrovicTT) [inline models/tt_transformers/tt/generator.py]: it seems that the whole logic which was working in both cases before (page_table == None and page_table != None) is now guarded for page_table != None. have you tested the other code-path?

File: models/tt_transformers/tt/generator.py (hunk 8 above)
Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py]: You are right, good catch, thanks!

File: models/tt_transformers/tt/generator.py (hunk 9 above)
Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py]: I think this is cleaner instead of calling it in every iteration.

File: models/tt_transformers/tt/generator.py (hunk 10 above)
Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py]: I'm not sure what you meant here exactly, I've refactored it a bit to avoid checking twice, let me know if it's ok now.

File: models/tt_transformers/tt/generator.py (hunk 9 above)
Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py]: Thanks for the reminder, I just ran `simple_text_demo.py` with `paged_attention=False` for `batch=32` in various configurations and it works properly. I'm not sure what you meant exactly by this, if `page_table` is not provided then `group_user_id` is calculated relative to a particular model instance, otherwise it's zero since only use it to read from the `page_table_user`.

File: models/tt_transformers/tt/generator.py (hunk 10 above)
Comment (ppetrovicTT) [reply models/tt_transformers/tt/generator.py]: sorry, i misread indentations, thinking that the whole block below is under "if page_table is not None"

File: models/tt_transformers/tt/generator.py [hunk 11]
```diff
@@ -56,66 +56,60 @@ def __init__(self, model, model_args, mesh_device, tokenizer=None, formatter=Non
...
-                    prefill_ids,
-                    page_table=page_table_user if page_table is not None else None,
-                    user_id=group_user_id,
-                    last_token_idx=last_token_idx,
-                    kv_cache=kv_cache[model_id] if kv_cache is not None else None,
-                    model_id=model_id,
-                )
-                out_list.append(logits)
+        for idx, user_id in enumerate(empty_slots):
+            logger.info(f"Prefilling User {user_id + 1}")
```
Comment (ppetrovicTT) [inline models/tt_transformers/tt/generator.py]: while here, can you please add seqlen info to this message?
it means A LOT while debugging failures in Shield CI

File: models/tt_transformers/tt/generator.py [hunk 12]
```diff
@@ -56,66 +56,60 @@ def __init__(self, model, model_args, mesh_device, tokenizer=None, formatter=Non
...
+            last_token_idx = seq_len - 1
 
-                user_id = group_user_id + model_id * max_batch_size_per_model
-                out = out_list[user_id]
+            prefill_seq_len = get_padded_prefill_len(seq_len)
+            prefill_ids = torch.cat(
+                [tokens[idx : idx + 1, :seq_len], torch.zeros(1, prefill_seq_len - seq_len).long()], dim=-1
+            )
+            page_table_user = (
+                self._get_prefill_user_page_table(page_table[idx : idx + 1], kv_cache[model_id], seq_len)
```
Comment (ppetrovicTT) [inline models/tt_transformers/tt/generator.py:91]: i think it would be great to add a block comment explaining this logic. i know it took you a while to understand how users were being iterated through in the previous implementation, and this would probably help a lot future debugging or improvements.

**Review by ppetrovicTT — APPROVED** [review]

File: models/tt_transformers/tt/generator.py (hunk 1 above)
Comment (skhorasganiTT) [reply models/tt_transformers/tt/generator.py:68]: It seems like this was removed again after being re-added

Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py:68]: Moved it to the top of the method

Comment (skhorasganiTT) [reply models/tt_transformers/tt/generator.py:68]: ah sorry, thx

File: models/tt_transformers/tt/generator.py [hunk 13]
```diff
@@ -56,66 +56,62 @@ def __init__(self, model, model_args, mesh_device, tokenizer=None, formatter=Non
...
-                user_id = group_user_id + model_id * max_batch_size_per_model
-                out = out_list[user_id]
+        for idx, user_id in enumerate(empty_slots):
+            model_id = user_id // max_batch_size_per_model
+            group_user_id = user_id % max_batch_size_per_model if page_table is None else 0
+            seq_len = int(prompt_lens[idx])
+            last_token_idx = seq_len - 1
+            prefill_seq_len = get_padded_prefill_len(seq_len)
+
+            logger.info(f"Prefilling User {user_id + 1} up to {prefill_seq_len} tokens")
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: Could you change this log to print the unpadded sequence length (i.e. `seq_len`)? I think it will cause confusion for users if they don't expect the padding, also for debugging I think we'll want to see what the original seq len was. In the "Finished prefill for ..." log below we also print the unpadded one.

Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py]: Done

**Review by skhorasganiTT — APPROVED** [review]
Thanks @rdraskicTT! Could you also post the vllm nightly run in the PR description?
